larpandoracontent/LArTwoDReco/LArCosmicRay/TrackExtensionRefinementAlgorithm.cc

```diff
@@ -0,0 +1,416 @@
+    unsigned int count(0);
+    CartesianVector extrapolatedEndPosition(0.f, 0.f, 0.f);
+    unsigned int hitsCollected(std::numeric_limits<int>::max());
+    const float slidingFitPitch(LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+    while (hitsCollected)
+    {
+        hitsCollected = 0;
+
+        try
+        {
+            const TwoDSlidingFitResult extrapolatedFit(&runningFitPositionVector, m_macroSlidingFitWindow, slidingFitPitch);
+            
+            if (count > 0)
+            {
+                extrapolatedStartPosition = clusterAssociation.IsEndUpstream() ? extrapolatedFit.GetGlobalMinLayerPosition() : extrapolatedFit.GetGlobalMaxLayerPosition();
+                extrapolatedDirection = clusterAssociation.IsEndUpstream() ? extrapolatedFit.GetGlobalMinLayerDirection() * (-1.f) : extrapolatedFit.GetGlobalMaxLayerDirection();
+            }
+            
+            extrapolatedEndPosition = extrapolatedStartPosition + (extrapolatedDirection * m_growingFitSegmentLength);
+
+            for (const Cluster *const pCluster : clustersInRegion)
+            {
+                for (const CaloHit *const pCaloHit : hitsInRegion.at(pCluster))
+                {
+                    // ATTN: To avoid counting same hit twice
+                    const ClusterToCaloHitListMap::iterator iter(clusterToCaloHitListMap.find(pCluster));
+                    if (iter != clusterToCaloHitListMap.end())
+                    {
+                        if (std::find(iter->second.begin(), iter->second.end(), pCaloHit) != iter->second.end())
+                            continue;
+                    }
+
+                    CartesianVector hitPosition(m_hitWidthMode ?
+                        LArHitWidthHelper::GetClosestPointToLine2D(extrapolatedStartPosition, extrapolatedDirection, pCaloHit) :
+                        pCaloHit->GetPositionVector());
+                    
+                    if (!IsInLineSegment(extrapolatedStartPosition, extrapolatedEndPosition, hitPosition))
+                        continue;
+
+                    const float transverseDistanceFromLine(extrapolatedDirection.GetCrossProduct(hitPosition - extrapolatedStartPosition).GetMagnitude());
+                    
+                    if (transverseDistanceFromLine > m_distanceToLine)
+                        continue;
+
+                    ++hitsCollected;
+                    
+                    runningFitPositionVector.push_back(hitPosition);
+                    clusterToCaloHitListMap[pCluster].push_back(pCaloHit);
+                }
+            }
+        }
+        catch (const StatusCodeException &)
+        {
+            return;
+        }
+
+        ++count;
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TrackExtensionRefinementAlgorithm::AreExtrapolatedHitsGood(const ClusterToCaloHitListMap &clusterToCaloHitListMap,
```

Comment:
As above, there are a lot of similarities between this function and the sibling in TrackMergeRefinementAlgorithm. Is clusterAssociation actually modified during the course of this algorithm? If not, it may be possible to do a little reorganisation of the order of this function and again move the bulk of the checks to the base class and then handle the specific cases in the inheriting classes (the additional variable needed for the higher boundary check may make this a little more cumbersome in this case).